File: quote-lambda-tf-backend/src/main/java/ebulter/quote/lambda/service/QuoteService.java
package ebulter.quote.lambda.service;

import ebulter.quote.lambda.client.ZenClient;
import ebulter.quote.lambda.model.Quote;
import ebulter.quote.lambda.model.UserLike;
import ebulter.quote.lambda.model.UserProgress;
import ebulter.quote.lambda.repository.QuoteRepository;
import ebulter.quote.lambda.repository.UserLikeRepository;
import ebulter.quote.lambda.repository.UserProgressRepository;
import ebulter.quote.lambda.util.QuoteUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

public class QuoteService {
    private static final Logger logger = LoggerFactory.getLogger(QuoteService.class);
    private final QuoteRepository quoteRepository;
    private final UserLikeRepository userLikeRepository;
    private UserProgressRepository userProgressRepository;

    public QuoteService(QuoteRepository quoteRepository, UserLikeRepository userLikeRepository) {
        this.quoteRepository = quoteRepository;
        this.userLikeRepository = userLikeRepository;
        try {
            this.userProgressRepository = new UserProgressRepository();
        } catch (Exception e) {
            logger.warn("Could not initialize UserProgressRepository, sequential navigation will be disabled: {}", e.getMessage());
            this.userProgressRepository = null;
        }
    }
    
    // Constructor for testing that allows injecting UserProgressRepository
    public QuoteService(QuoteRepository quoteRepository, UserLikeRepository userLikeRepository, UserProgressRepository userProgressRepository) {
        this.quoteRepository = quoteRepository;
        this.userLikeRepository = userLikeRepository;
        this.userProgressRepository = userProgressRepository;
    }

    public Quote getQuote(String username, final Set<Integer> idsToExclude) {
        // For authenticated users, use sequential navigation if available
        if (username != null && !username.isEmpty()) {
            if (userProgressRepository != null) {
                return getNextSequentialQuote(username);
            } else {
                // Fallback to old behavior if UserProgressRepository is not available
                return getRandomQuoteForUnauthenticatedUser(idsToExclude);
            }
        }
        
        // For unauthenticated users, use the old random approach (but exclude provided IDs)
        return getRandomQuoteForUnauthenticatedUser(idsToExclude);
    }

    /**
     * Get the next sequential quote for an authenticated user
     */
    private Quote getNextSequentialQuote(String username) {
        logger.info("Getting next sequential quote for user: {}", username);
        
        // Get user's current progress
        UserProgress userProgress = userProgressRepository.getUserProgress(username);
        int nextQuoteId;
        
        if (userProgress == null) {
            // New user - start with quote ID 1
            nextQuoteId = 1;
            logger.info("New user {} starting with quote ID: {}", username, nextQuoteId);
        } else {
            // Existing user - get next quote
            nextQuoteId = userProgress.getLastQuoteId() + 1;
            logger.info("User {} progress: lastQuoteId={}, nextQuoteId={}", username, userProgress.getLastQuoteId(), nextQuoteId);
        }
        
        // Check if we need to fetch more quotes from Zen
        int maxId = quoteRepository.getMaxQuoteId();
        if (nextQuoteId > maxId) {
            logger.info("Next quote ID {} exceeds max ID {}, fetching more quotes", nextQuoteId, maxId);
            fetchMoreQuotesIfNeeded();
            maxId = quoteRepository.getMaxQuoteId();
        }
        
        // Get the quote
        Quote quote = quoteRepository.findById(nextQuoteId);
        if (quote == null) {
            logger.warn("Quote with ID {} not found, finding next available quote", nextQuoteId);
            quote = findNextAvailableQuote(nextQuoteId);
        }
        
        if (quote != null) {
            // Update user progress
            userProgressRepository.updateLastQuoteId(username, quote.getId());
            logger.info("Updated user {} progress to lastQuoteId={}", username, quote.getId());
        }
        
        return quote;
    }

    /**
     * Get random quote for unauthenticated users (legacy behavior)
     */
    private Quote getRandomQuoteForUnauthenticatedUser(Set<Integer> idsToExclude) {
        logger.info("Getting random quote for unauthenticated user, excluding {} IDs", idsToExclude.size());
        
        int maxId = quoteRepository.getMaxQuoteId();
        logger.info("Max quote ID in database: {}", maxId);
        
        // Check if we need to fetch more quotes from Zen
        if (maxId < 5 || maxId <= idsToExclude.size()) {
            fetchMoreQuotesIfNeeded();
            maxId = quoteRepository.getMaxQuoteId();
        }
        
        // Try to find a valid quote by checking random IDs
        Random random = new Random();
        int maxAttempts = Math.min(100, maxId);
        Set<Integer> attemptedIds = new HashSet<>();
        
        for (int attempt = 0; attempt < maxAttempts; attempt++) {
            int candidateId = random.nextInt(maxId) + 1;
            
            if (idsToExclude.contains(candidateId) || attemptedIds.contains(candidateId)) {
                continue;
            }
            
            attemptedIds.add(candidateId);
            
            Quote quote = quoteRepository.findById(candidateId);
            if (quote != null) {
                logger.info("Selected random quote ID: {} after {} attempts", candidateId, attempt + 1);
                return quote;
            }
        }
        
        // Fallback to full scan
        logger.warn("Could not find valid quote with {} random attempts, falling back to full scan", maxAttempts);
        List<Quote> allQuotes = quoteRepository.getAllQuotes();
        List<Quote> filteredQuotes = allQuotes.stream()
                .filter(quote -> !idsToExclude.contains(quote.getId()))
                .toList();
        
        if (filteredQuotes.isEmpty()) {
            logger.warn("No available quotes after excluding {} IDs", idsToExclude.size());
            return null;
        }
        
        int randomIndex = random.nextInt(filteredQuotes.size());
        Quote selectedQuote = filteredQuotes.get(randomIndex);
        logger.info("Selected quote ID: {} using fallback method", selectedQuote.getId());
        return selectedQuote;
    }

    /**
     * Fetch more quotes from Zen if needed
     */
    private void fetchMoreQuotesIfNeeded() {
        try {
            logger.info("start fetching quotes from Zen");
            Set<Quote> fetchedQuotes = ZenClient.getSomeUniqueQuotes();
            logger.info("finished fetching quotes from Zen, fetched {} quotes", fetchedQuotes.size());
            
            List<Quote> currentDatabaseQuotes = quoteRepository.getAllQuotes();
            fetchedQuotes.removeAll(new HashSet<>(currentDatabaseQuotes));
            AtomicInteger idGenerator = new AtomicInteger(currentDatabaseQuotes.size() + 1);
            fetchedQuotes.forEach(quote -> quote.setId(idGenerator.getAndIncrement()));
            logger.info("start saving {} quotes", fetchedQuotes.size());
            quoteRepository.saveAll(fetchedQuotes);
            logger.info("finished saving {} quotes", fetchedQuotes.size());
        } catch (IOException e) {
            logger.error("Failed to read quotes from ZenQuotes: {}", e.getMessage());
        }
    }

    /**
     * Find the next available quote ID (handles gaps in sequence)
     */
    private Quote findNextAvailableQuote(int startId) {
        int maxId = quoteRepository.getMaxQuoteId();
        for (int id = startId; id <= maxId; id++) {
            Quote quote = quoteRepository.findById(id);
            if (quote != null) {
                logger.info("Found next available quote ID: {}", id);
                return quote;
            }
        }
        return null;
    }

    public Quote likeQuote(String username, int quoteId) {
        Quote quote = quoteRepository.findById(quoteId);
        if (quote != null) {
            // Get max order for user and set new order to max + 1
            int maxOrder = userLikeRepository.getMaxOrderForUser(username);
            int newOrder = maxOrder + 1;
            
            UserLike userLike = new UserLike(username, quoteId, System.currentTimeMillis(), newOrder);
            userLikeRepository.saveUserLike(userLike);
            return quote;
        } else {
            return QuoteUtil.getErrorQuote("Quote to like not found");
        }
    }

    public void unlikeQuote(String username, int quoteId) {
        userLikeRepository.deleteUserLike(username, quoteId);
    }

    public List<Quote> getLikedQuotes() {
        // Get all quotes that have at least one like
        List<Quote> allQuotes = quoteRepository.getAllQuotes();
        return allQuotes.stream()
                .filter(quote -> userLikeRepository.getLikeCount(quote.getId()) > 0)
                .sorted((q1, q2) -> q1.getId() - q2.getId())
                .toList();
    }
    
    public List<Quote> getLikedQuotesByUser(String username) {
        List<UserLike> userLikes = userLikeRepository.getLikesByUser(username);
        // userLikes are already sorted by order from repository
        return userLikes.stream()
                .map(userLike -> quoteRepository.findById(userLike.getQuoteId()))
                .filter(quote -> quote != null)
                .toList();
    }

    public int getLikeCount(int quoteId) {
        return userLikeRepository.getLikeCount(quoteId);
    }

    public boolean hasUserLikedQuote(String username, int quoteId) {
        return userLikeRepository.hasUserLikedQuote(username, quoteId);
    }

    /**
     * Get quote by ID for sequential navigation
     */
    public Quote getQuoteById(String username, int quoteId) {
        logger.info("Getting quote {} for user {}", quoteId, username);
        Quote quote = quoteRepository.findById(quoteId);
        
        if (quote != null && username != null && !username.isEmpty() && userProgressRepository != null) {
            // Update user progress to this quote ID
            userProgressRepository.updateLastQuoteId(username, quoteId);
            logger.info("Updated user {} progress to lastQuoteId={}", username, quoteId);
        }
        
        return quote;
    }

    /**
     * Get previous quote for sequential navigation
     */
    public Quote getPreviousQuote(String username, int currentQuoteId) {
        if (username == null || username.isEmpty()) {
            return null;
        }
        
        logger.info("Getting previous quote for user {} from current ID {}", username, currentQuoteId);
        
        // Find previous available quote
        for (int id = currentQuoteId - 1; id >= 1; id--) {
            Quote quote = quoteRepository.findById(id);
            if (quote != null) {
                // Update user progress
                userProgressRepository.updateLastQuoteId(username, id);
                logger.info("Updated user {} progress to previous quote ID {}", username, id);
                return quote;
            }
        }
        
        logger.info("No previous quote available for user {}", username);
        return null;
    }

    /**
     * Get next quote for sequential navigation
     */
    public Quote getNextQuote(String username, int currentQuoteId) {
        if (username == null || username.isEmpty()) {
            return null;
        }
        
        logger.info("Getting next quote for user {} from current ID {}", username, currentQuoteId);
        
        // Find next available quote
        int maxId = quoteRepository.getMaxQuoteId();
        for (int id = currentQuoteId + 1; id <= maxId; id++) {
            Quote quote = quoteRepository.findById(id);
            if (quote != null) {
                // Update user progress
                userProgressRepository.updateLastQuoteId(username, id);
                logger.info("Updated user {} progress to next quote ID {}", username, id);
                return quote;
            }
        }
        
        logger.info("No next quote available for user {}", username);
        return null;
    }

    /**
     * Get user's current progress (last quote ID)
     */
    public UserProgress getUserProgress(String username) {
        if (username == null || username.isEmpty() || userProgressRepository == null) {
            return null;
        }
        return userProgressRepository.getUserProgress(username);
    }

    /**
     * Get all quotes from 1 to lastQuoteId for Viewed Quotes screen
     */
    public List<Quote> getViewedQuotes(String username) {
        if (username == null || username.isEmpty() || userProgressRepository == null) {
            return List.of();
        }
        
        UserProgress userProgress = userProgressRepository.getUserProgress(username);
        if (userProgress == null) {
            return List.of();
        }
        
        int lastQuoteId = userProgress.getLastQuoteId();
        logger.info("Getting quotes 1 to {} for user {}", lastQuoteId, username);
        
        List<Quote> viewedQuotes = new ArrayList<>();
        for (int id = 1; id <= lastQuoteId; id++) {
            Quote quote = quoteRepository.findById(id);
            if (quote != null) {
                viewedQuotes.add(quote);
            }
        }
        
        logger.info("Found {} quotes for user {}", viewedQuotes.size(), username);
        return viewedQuotes;
    }

    public void recordView(String username, int quoteId) {
        if (username != null && !username.isEmpty()) {
            if (userProgressRepository != null) {
                // Update user progress instead of recording individual views
                UserProgress currentProgress = userProgressRepository.getUserProgress(username);
                if (currentProgress == null) {
                    currentProgress = new UserProgress(username, quoteId, System.currentTimeMillis());
                } else {
                    currentProgress.setLastQuoteId(quoteId);
                    currentProgress.setUpdatedAt(System.currentTimeMillis());
                }
                userProgressRepository.saveUserProgress(currentProgress);
                logger.info("Updated user progress for user {} to lastQuoteId {}", username, quoteId);
            } else {
                logger.warn("Cannot record view: UserProgressRepository is not initialized");
            }
        }
    }

    public List<Quote> getViewedQuotesByUser(String username) {
        // Return quotes 1 to lastQuoteId using the new sequential system
        if (userProgressRepository == null) {
            logger.warn("Cannot get viewed quotes: UserProgressRepository is not initialized");
            return new ArrayList<>();
        }
        
        UserProgress progress = userProgressRepository.getUserProgress(username);
        if (progress == null || progress.getLastQuoteId() <= 0) {
            return new ArrayList<>();
        }
        
        List<Quote> viewedQuotes = new ArrayList<>();
        for (int i = 1; i <= progress.getLastQuoteId(); i++) {
            Quote quote = quoteRepository.findById(i);
            if (quote != null) {
                viewedQuotes.add(quote);
            }
        }
        return viewedQuotes;
    }

    /**
     * Reorder a liked quote for a user
     * Updates all affected likes to maintain sequential ordering
     */
    public void reorderLikedQuote(String username, int quoteId, int newOrder) {
        List<UserLike> allLikes = userLikeRepository.getLikesByUser(username);
        
        // Find the like to move
        UserLike likeToMove = allLikes.stream()
                .filter(like -> like.getQuoteId() == quoteId)
                .findFirst()
                .orElse(null);
        
        if (likeToMove == null) {
            logger.warn("Quote {} not found in user {}'s likes", quoteId, username);
            return;
        }
        
        int oldOrder = likeToMove.getOrder() != null ? likeToMove.getOrder() : allLikes.indexOf(likeToMove) + 1;
        
        if (oldOrder == newOrder) {
            return; // No change needed
        }
        
        // Update orders for affected likes
        if (newOrder > oldOrder) {
            // Moving down: decrement orders between oldOrder and newOrder
            allLikes.stream()
                    .filter(like -> like.getOrder() != null && like.getOrder() > oldOrder && like.getOrder() <= newOrder)
                    .forEach(like -> {
                        like.setOrder(like.getOrder() - 1);
                        userLikeRepository.saveUserLike(like);
                    });
        } else {
            // Moving up: increment orders between newOrder and oldOrder
            allLikes.stream()
                    .filter(like -> like.getOrder() != null && like.getOrder() >= newOrder && like.getOrder() < oldOrder)
                    .forEach(like -> {
                        like.setOrder(like.getOrder() + 1);
                        userLikeRepository.saveUserLike(like);
                    });
        }
        
        // Set the moved item to new order
        likeToMove.setOrder(newOrder);
        userLikeRepository.saveUserLike(likeToMove);
        
        logger.info("Reordered quote {} for user {} from order {} to {}", quoteId, username, oldOrder, newOrder);
    }

}

File: quote-azure-backend/src/Services/QuoteService.cs
using Microsoft.Extensions.Logging;
using QuoteAzureBackend.Models;
using QuoteAzureBackend.Data;

namespace QuoteAzureBackend.Services
{
    public interface IQuoteService
    {
        Task<Quote?> GetQuoteAsync(string? username, HashSet<int> idsToExclude);
        Task<Quote?> GetQuoteByIdAsync(string? username, int quoteId);
        Task<Quote?> LikeQuoteAsync(string username, int quoteId);
        Task UnlikeQuoteAsync(string username, int quoteId);
        Task<List<Quote>> GetLikedQuotesByUserAsync(string username);
        Task<int> GetLikeCountAsync(int quoteId);
        Task<bool> HasUserLikedQuoteAsync(string username, int quoteId);
        Task<Quote?> GetPreviousQuoteAsync(string username, int currentQuoteId);
        Task<Quote?> GetNextQuoteAsync(string username, int currentQuoteId);
        Task<UserProgress?> GetUserProgressAsync(string username);
        Task<List<Quote>> GetViewedQuotesAsync(string username);
        Task<bool> RecordViewAsync(string username, int quoteId);
        Task ReorderLikedQuoteAsync(string username, int quoteId, int newOrder);
        Task ResetUserProgressAsync(string username);
    }

    public class QuoteService : IQuoteService
    {
        private readonly IQuoteRepository _quoteRepository;
        private readonly IUserActivityRepository _userActivityRepository;
        private readonly IZenQuotesService _zenQuotesService;
        private readonly ILogger<QuoteService> _logger;

        public QuoteService(
            IQuoteRepository quoteRepository,
            IUserActivityRepository userActivityRepository,
            IZenQuotesService zenQuotesService,
            ILogger<QuoteService> logger)
        {
            _quoteRepository = quoteRepository;
            _userActivityRepository = userActivityRepository;
            _zenQuotesService = zenQuotesService;
            _logger = logger;
        }

        public async Task<Quote?> GetQuoteAsync(string? username, HashSet<int> idsToExclude)
        {
            if (!string.IsNullOrEmpty(username))
            {
                return await GetNextSequentialQuoteAsync(username);
            }
            return await GetRandomQuoteForUnauthenticatedUserAsync(idsToExclude);
        }

        private async Task<Quote?> GetNextSequentialQuoteAsync(string username)
        {
            _logger.LogInformation("Getting next sequential quote for user: {Username}", username);
            
            // Get user's current progress (matching Java implementation)
            var userProgress = await _userActivityRepository.GetUserProgressAsync(username);
            int nextQuoteId;
            
            if (userProgress == null)
            {
                // New user - start with quote ID 1
                nextQuoteId = 1;
                _logger.LogInformation("New user {Username} starting with quote ID: {NextQuoteId}", username, nextQuoteId);
            }
            else
            {
                // Existing user - get next quote
                nextQuoteId = userProgress.LastQuoteId + 1;
                _logger.LogInformation("User {Username} progress: lastQuoteId={LastQuoteId}, nextQuoteId={NextQuoteId}", 
                    username, userProgress.LastQuoteId, nextQuoteId);
            }
            
            // Check if we need to fetch more quotes
            var allQuotes = await _quoteRepository.GetAllQuotesAsync();
            int maxId = allQuotes.Any() ? allQuotes.Max(q => q.Id) : 0;
            
            if (nextQuoteId > maxId)
            {
                _logger.LogInformation("Next quote ID {NextQuoteId} exceeds max ID {MaxId}, fetching more quotes", 
                    nextQuoteId, maxId);
                await FetchMoreQuotesIfNeededAsync();
                allQuotes = await _quoteRepository.GetAllQuotesAsync();
                maxId = allQuotes.Any() ? allQuotes.Max(q => q.Id) : 0;
            }
            
            // Get the quote
            var quote = await _quoteRepository.GetQuoteByIdAsync(nextQuoteId);
            if (quote == null)
            {
                _logger.LogWarning("Quote with ID {NextQuoteId} not found, finding next available quote", nextQuoteId);
                quote = await FindNextAvailableQuoteAsync(nextQuoteId);
            }
            
            if (quote != null)
            {
                // Update user progress (matching Java implementation)
                await _userActivityRepository.UpdateLastQuoteIdAsync(username, quote.Id);
                _logger.LogInformation("Updated user {Username} progress to lastQuoteId={LastQuoteId}", 
                    username, quote.Id);
            }
            
            return quote;
        }

        private async Task<Quote?> GetRandomQuoteForUnauthenticatedUserAsync(HashSet<int> idsToExclude)
        {
            _logger.LogInformation("Getting random quote for unauthenticated user, excluding {Count} IDs", idsToExclude.Count);
            
            var allQuotes = await _quoteRepository.GetAllQuotesAsync();
            int maxId = allQuotes.Any() ? allQuotes.Max(q => q.Id) : 0;
            _logger.LogInformation("Max quote ID in database: {MaxId}", maxId);
            
            if (maxId < 5 || maxId <= idsToExclude.Count)
            {
                _logger.LogInformation("Need to fetch more quotes (maxId={MaxId}, excludeCount={ExcludeCount})", maxId, idsToExclude.Count);
                await FetchMoreQuotesIfNeededAsync();
                allQuotes = await _quoteRepository.GetAllQuotesAsync();
                maxId = allQuotes.Any() ? allQuotes.Max(q => q.Id) : 0;
                _logger.LogInformation("After fetching, new max ID: {MaxId}", maxId);
            }
            
            var random = new Random();
            int maxAttempts = Math.Min(100, maxId);
            var attemptedIds = new HashSet<int>();
            
            for (int attempt = 0; attempt < maxAttempts; attempt++)
            {
                int candidateId = random.Next(maxId) + 1;
                if (idsToExclude.Contains(candidateId) || attemptedIds.Contains(candidateId))
                    continue;
                
                attemptedIds.Add(candidateId);
                var quote = await _quoteRepository.GetQuoteByIdAsync(candidateId);
                if (quote != null)
                    return quote;
            }
            
            var filteredQuotes = allQuotes.Where(q => !idsToExclude.Contains(q.Id)).ToList();
            if (!filteredQuotes.Any())
            {
                _logger.LogWarning("No available quotes after excluding {Count} IDs", idsToExclude.Count);
                return null;
            }
            
            return filteredQuotes[random.Next(filteredQuotes.Count)];
        }

        private async Task FetchMoreQuotesIfNeededAsync()
        {
            try
            {
                _logger.LogInformation("Fetching quotes from ZenQuotes API");
                var fetchedQuotes = await _zenQuotesService.GetMultipleQuotesAsync(50);
                _logger.LogInformation("Fetched {Count} quotes from ZenQuotes", fetchedQuotes?.Count ?? 0);
                
                var currentDatabaseQuotes = await _quoteRepository.GetAllQuotesAsync();
                _logger.LogInformation("Current database has {Count} quotes", currentDatabaseQuotes.Count);
                
                var existingTexts = new HashSet<string>(currentDatabaseQuotes.Select(q => q.QuoteText));
                int nextId = currentDatabaseQuotes.Any() ? currentDatabaseQuotes.Max(q => q.Id) + 1 : 1;
                
                int addedCount = 0;
                if (fetchedQuotes != null)
                {
                    foreach (var quote in fetchedQuotes)
                {
                    if (!existingTexts.Contains(quote.QuoteText))
                    {
                        var originalId = quote.Id;
                        quote.Id = nextId++;
                        _logger.LogInformation("Assigning new ID: {NewId} to quote (original ID: {OriginalId})", quote.Id, originalId);
                        await _quoteRepository.AddQuoteAsync(quote);
                        existingTexts.Add(quote.QuoteText);
                        addedCount++;
                    }
                }
                }
                
                _logger.LogInformation("Added {Count} new quotes to database", addedCount);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to fetch quotes from ZenQuotes");
            }
        }

        private async Task<Quote?> FindNextAvailableQuoteAsync(int startId)
        {
            var allQuotes = await _quoteRepository.GetAllQuotesAsync();
            int maxId = allQuotes.Any() ? allQuotes.Max(q => q.Id) : 0;
            
            for (int id = startId; id <= maxId; id++)
            {
                var quote = await _quoteRepository.GetQuoteByIdAsync(id);
                if (quote != null)
                    return quote;
            }
            return null;
        }

        public async Task<Quote?> GetQuoteByIdAsync(string? username, int quoteId)
        {
            var quote = await _quoteRepository.GetQuoteByIdAsync(quoteId);
            if (quote != null && !string.IsNullOrEmpty(username))
            {
                await _userActivityRepository.UpdateUserPreferencesAsync(new UserPreferences
                {
                    UserId = username,
                    LastQuoteId = quoteId,
                    UpdatedAt = DateTime.UtcNow
                });
            }
            return quote;
        }

        public async Task<Quote?> LikeQuoteAsync(string username, int quoteId)
        {
            var quote = await _quoteRepository.GetQuoteByIdAsync(quoteId);
            if (quote != null)
            {
                await _userActivityRepository.AddUserLikeAsync(username, quoteId);
            }
            return quote;
        }

        public async Task UnlikeQuoteAsync(string username, int quoteId)
        {
            await _userActivityRepository.RemoveUserLikeAsync(username, quoteId);
        }

        public async Task<List<Quote>> GetLikedQuotesByUserAsync(string username)
        {
            var allLikes = await _userActivityRepository.GetAllUserLikesAsync(username);
            var likedQuotes = new List<Quote>();
            foreach (var like in allLikes.OrderBy(l => l.Order))
            {
                var quote = await _quoteRepository.GetQuoteByIdAsync(like.QuoteId);
                if (quote != null)
                    likedQuotes.Add(quote);
            }
            return likedQuotes;
        }

        public Task<int> GetLikeCountAsync(int quoteId) => Task.FromResult(0);

        public async Task<bool> HasUserLikedQuoteAsync(string username, int quoteId)
        {
            var likedIds = await _userActivityRepository.GetUserLikedQuoteIdsAsync(username);
            return likedIds.Contains(quoteId);
        }

        public async Task<Quote?> GetPreviousQuoteAsync(string username, int currentQuoteId)
        {
            if (string.IsNullOrEmpty(username)) return null;
            
            for (int id = currentQuoteId - 1; id >= 1; id--)
            {
                var quote = await _quoteRepository.GetQuoteByIdAsync(id);
                if (quote != null)
                {
                    await _userActivityRepository.UpdateUserPreferencesAsync(new UserPreferences
                    {
                        UserId = username,
                        LastQuoteId = id,
                        UpdatedAt = DateTime.UtcNow
                    });
                    return quote;
                }
            }
            return null;
        }

        public async Task<Quote?> GetNextQuoteAsync(string username, int currentQuoteId)
        {
            if (string.IsNullOrEmpty(username)) return null;
            
            var allQuotes = await _quoteRepository.GetAllQuotesAsync();
            int maxId = allQuotes.Any() ? allQuotes.Max(q => q.Id) : 0;
            
            for (int id = currentQuoteId + 1; id <= maxId; id++)
            {
                var quote = await _quoteRepository.GetQuoteByIdAsync(id);
                if (quote != null)
                {
                    await _userActivityRepository.UpdateUserPreferencesAsync(new UserPreferences
                    {
                        UserId = username,
                        LastQuoteId = id,
                        UpdatedAt = DateTime.UtcNow
                    });
                    return quote;
                }
            }
            return null;
        }

        public async Task<UserProgress?> GetUserProgressAsync(string username)
        {
            if (string.IsNullOrEmpty(username)) return null;
            var preferences = await _userActivityRepository.GetUserPreferencesAsync(username);
            if (preferences == null) return null;
            return new UserProgress
            {
                Username = username,
                LastQuoteId = preferences.LastQuoteId,
                UpdatedAt = preferences.UpdatedAt ?? DateTime.UtcNow
            };
        }

        public async Task<List<Quote>> GetViewedQuotesAsync(string username)
        {
            // Return quotes 1 to lastQuoteId using the sequential system (matching Java implementation)
            _logger.LogInformation("Getting viewed quotes for user: {Username}", username);
            
            if (string.IsNullOrEmpty(username))
            {
                return new List<Quote>();
            }
            
            var progress = await _userActivityRepository.GetUserProgressAsync(username);
            if (progress == null || progress.LastQuoteId <= 0)
            {
                _logger.LogInformation("User {Username} has no progress or hasn't viewed any quotes", username);
                return new List<Quote>();
            }
            
            var viewedQuotes = new List<Quote>();
            for (int i = 1; i <= progress.LastQuoteId; i++)
            {
                var quote = await _quoteRepository.GetQuoteByIdAsync(i);
                if (quote != null)
                {
                    viewedQuotes.Add(quote);
                }
                else
                {
                    _logger.LogWarning("Quote with ID {QuoteId} not found while getting viewed quotes for user {Username}", i, username);
                }
            }
            
            _logger.LogInformation("Retrieved {Count} viewed quotes for user {Username}", viewedQuotes.Count, username);
            return viewedQuotes;
        }

        public Task<bool> RecordViewAsync(string userId, int quoteId)
        {
            try
            {
                // In the Java version, this updates the user progress
                // But in C#, the progress is already updated in GetNextSequentialQuoteAsync
                // So we don't need to do anything here, but we keep the method for compatibility
                _logger.LogDebug("RecordViewAsync called for user {UserId}, quote {QuoteId} - progress already tracked", userId, quoteId);
                return Task.FromResult(true);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error recording view for user {UserId}, quote {QuoteId}", userId, quoteId);
                return Task.FromResult(false);
            }
        }

        
        public async Task ReorderLikedQuoteAsync(string username, int quoteId, int newOrder)
        {
            var allLikes = await _userActivityRepository.GetAllUserLikesAsync(username);
            
            // Find the like to move
            var likeToMove = allLikes.FirstOrDefault(l => l.QuoteId == quoteId);
            
            if (likeToMove == null)
            {
                _logger.LogWarning("Quote {QuoteId} not found in user {Username}'s likes", quoteId, username);
                return;
            }
            
            int oldOrder = likeToMove.Order > 0 ? likeToMove.Order : allLikes.IndexOf(likeToMove) + 1;
            
            if (oldOrder == newOrder)
            {
                return; // No change needed
            }
            
            // Update orders for affected likes
            if (newOrder > oldOrder)
            {
                // Moving down: decrement orders between oldOrder and newOrder
                var likesToUpdate = allLikes.Where(l => l.Order > oldOrder && l.Order <= newOrder);
                foreach (var like in likesToUpdate)
                {
                    await _userActivityRepository.UpdateUserLikeOrderAsync(username, like.QuoteId, like.Order - 1);
                }
            }
            else
            {
                // Moving up: increment orders between newOrder and oldOrder
                var likesToUpdate = allLikes.Where(l => l.Order >= newOrder && l.Order < oldOrder);
                foreach (var like in likesToUpdate)
                {
                    await _userActivityRepository.UpdateUserLikeOrderAsync(username, like.QuoteId, like.Order + 1);
                }
            }
            
            // Set the moved item to new order
            await _userActivityRepository.UpdateUserLikeOrderAsync(username, quoteId, newOrder);
            
            _logger.LogInformation("Reordered quote {QuoteId} for user {Username} from order {OldOrder} to {NewOrder}", 
                quoteId, username, oldOrder, newOrder);
        }

        public async Task ResetUserProgressAsync(string username)
        {
            if (string.IsNullOrEmpty(username)) return;
            await _userActivityRepository.UpdateUserPreferencesAsync(new UserPreferences
            {
                UserId = username,
                LastQuoteId = 0,
                UpdatedAt = DateTime.UtcNow
            });
        }
    }
}
